docs/rfc/004-integer-type-casts.md

```diff
@@ -0,0 +1,231 @@
+x + - y as u8
+```
+is like
+```
+x + ((- y) as u8)
+```
+This precedence is the same as in Rust:
+see [here](https://doc.rust-lang.org/stable/reference/expressions.html#expression-precedence).
+
+## Dynamic Semantics
+
+When the mathematical integer value of the expression
+is representable in the type that the expression is cast to,
+there is no question that the cast must succeed
+and merely change the type of the Leo value,
+but not its mathematical integer value.
+This is always the case when the cast is to a type
+with the same signedness and with the same or larger size.
+This is sometimes the case when
+the cast is to a type whose range does not cover the range of the source type
+but the value in question is in the intersection of the two ranges.
+
+When the mathematical integer value of the expression
+is not representable in the type that the expression is cast to,
+there are two possible approaches, discussed below.
+
+### Value-Preserving Casts
+
+The first approach is to deem that situation erroneous.
+That is, to require casts to always preserve the mathematical integer values.
+
+In this approach, casts only serve to change types, never values.
+When values are to be changed, separate (built-in) functions can be used,
+e.g. to mask bits and achieve the same effect as
+the value-changing casts discussed below.
+
+From a point of view, this approach seems to match Leo's
+treatment of potentially erroneous situations like integer overflows:
+the principle is that developers should explicitly use
+operations that may overflow if that is their intention,
+rather than having those situation possibly occur unexpectedly.
+
+A value-preserving cast to a type
+whose range does not cover the original type's range
+implicitly expresses a developer expectation that the value
+is actually in the intersection of the two types' ranges,
+in the same way that the use of integer addition
+implicitly expresses the expectation that the addition does not overflow.
+
+Consider this somewhat abstract example:
+```
+... // some computations on u32 values, which could not be done with u16
+let r: u32 = ...; // this is the final result of the u32 operations above
+let s: u16 = r as u16; // but r is expected to fit in u16, so we cast it here
+```
+With value-preserving casts, the expectation mentioned above
+is checked by the Leo compiler during proof generation,
+in the same way as with integer overflow.
+
+In the example above,
+if instead the variable `s` is meant to contain the low 16 bits of `r`,
+e.g. in a cryptographic computation,
+then the value-preserving cast should be preceded by
+an explicit operation to obtain the low 16 bits, making the intent clear:
+```
+... // some computations on u32 values, which could not be done with u16
+let r: u32 = ...; // this is the final result of the u32 operations above
+let r_low16: u32 = r & 0xFFFF; // assuming we have bitwise ops and hex literals
+let s: u16 = r_low16 as u16; // no value change here
+```
+
```

Comment:
The `value-preserving casts` alternative does not prevent the later addition of other operators that change the value and type, such as the re-interpretation between signed and unsigned.  It might improve clarity of code to have a syntactic difference between value-preserving and value-changing casts.

Reply:
I agree.